mono/metadata/threads.c

```diff
@@ -5065,15 +5088,67 @@ threads_add_joinable_thread_nolock (gpointer tid)
 }
 #endif
 
+static void
+threads_add_pending_joinable_runtime_thread (gpointer thread_info)
+{
+	g_assert (thread_info);
+	MonoThreadInfo *mono_thread_info = (MonoThreadInfo*)thread_info;
+
+	if (mono_thread_info->runtime_thread) {
+		gint32 old_joinable_state = mono_atomic_load_i32 (&mono_thread_info->thread_joinable_state);
+		if (old_joinable_state == THREAD_JOINABLE_STATE_NONE) {
+			if (mono_atomic_cas_i32 (&mono_thread_info->thread_joinable_state, THREAD_JOINABLE_STATE_PENDING, THREAD_JOINABLE_STATE_NONE) == THREAD_JOINABLE_STATE_NONE) {
+				gint32 result = mono_atomic_inc_i32 (&pending_joinable_runtime_thread_count);
+				g_assert (result > 0);
+			}
+		}
+	}
+}
+
+static gboolean
+threads_wait_pending_joinable_runtime_threads (uint32_t timeout)
+{
+	gboolean wait_result = TRUE;
+	if (mono_atomic_load_i32 (&pending_joinable_runtime_thread_count) > 0) {
+		joinable_threads_lock ();
+		if (timeout == MONO_INFINITE_WAIT) {
+			mono_os_cond_wait (&zero_pending_joinable_runtime_thread_event, &joinable_threads_mutex);
+		} else {
+			gint64 start = mono_msec_ticks ();
+			gint64 elapsed = 0;
+			while (mono_atomic_load_i32 (&pending_joinable_runtime_thread_count) > 0 && elapsed < timeout) {
+				wait_result = (mono_os_cond_timedwait (&zero_pending_joinable_runtime_thread_event, &joinable_threads_mutex, timeout - (uint32_t)elapsed) == 0);
+				elapsed = mono_msec_ticks () - start;
+			}
+		}
+		joinable_threads_unlock ();
+	}
+
+	return wait_result;
+}
+
 void
 mono_threads_add_joinable_runtime_thread (gpointer thread_info)
 {
 	g_assert (thread_info);
 	MonoThreadInfo *mono_thread_info = (MonoThreadInfo*)thread_info;
 
 	if (mono_thread_info->runtime_thread) {
-		if (mono_atomic_cas_i32 (&mono_thread_info->thread_pending_native_join, TRUE, FALSE) == FALSE)
-			mono_threads_add_joinable_thread ((gpointer)(MONO_UINT_TO_NATIVE_THREAD_ID (mono_thread_info_get_tid (mono_thread_info))));
+		gint32 old_joinable_state = mono_atomic_load_i32 (&mono_thread_info->thread_joinable_state);
+		if (old_joinable_state != THREAD_JOINABLE_STATE_ADDED) {
+			if (mono_atomic_cas_i32 (&mono_thread_info->thread_joinable_state, THREAD_JOINABLE_STATE_ADDED, old_joinable_state) == old_joinable_state) {
+				mono_threads_add_joinable_thread ((gpointer)(MONO_UINT_TO_NATIVE_THREAD_ID (mono_thread_info_get_tid (mono_thread_info))));
+				if (old_joinable_state == THREAD_JOINABLE_STATE_PENDING) {
+					gint32 result = mono_atomic_dec_i32 (&pending_joinable_runtime_thread_count);
+					g_assert (result >= 0);
+					if (result == 0) {
+						joinable_threads_lock ();
+						mono_os_cond_broadcast (&zero_pending_joinable_runtime_thread_event);
```

Comment:
How many waiters are we expecting for this cond var?

Reply:
In theory, more than 1, that's why I used mono_os_cond_broadcast to not risk leaving threads blocked on this wait, if it happens. Also note that it should be quite rare to actually have anyone waiting on this event during shutdown, then we are getting into the race condition this PR solves, that happens, but not very frequently.